File: Solarica/connector/dotnet/SolaricaConnector/Services/SerialDriver.cs
using System.IO.Ports;
using SolaricaConnector.Models;

namespace SolaricaConnector.Services;

/// <summary>
/// Reads measurements directly from a PVPM 1540X over a serial/USB-CDC port.
///
/// Note: The PVPM serial protocol is not yet reverse-engineered.
/// This driver correctly enumerates COM ports; FetchMeasurements() throws
/// NotImplementedException until the protocol is documented.
/// </summary>
public sealed class SerialDriver : IDeviceDriver
{
    private bool _connected;
    private string? _port;
    private string? _lastError;

    public DeviceStatus Detect() => new(
        Connected: _connected,
        Mode: "serial",
        Port: _port,
        DeviceModel: null,
        DeviceSerial: null,
        FirmwareVersion: null,
        TransferModeRequired: true,
        TransferModeDetected: false,
        LastError: _lastError
    );

    public IReadOnlyList<PortInfo> ListPorts() =>
        SerialPort.GetPortNames()
            .Select(name => new PortInfo(name, $"Serial port {name}"))
            .ToList();

    public DeviceStatus Connect(string port)
    {
        _connected = true;
        _port = port;
        _lastError =
            "Serial protocol not implemented. " +
            "Use the vendor_export driver to read exported files, " +
            "or capture serial traffic to implement this driver.";
        return Detect();
    }

    public void Disconnect()
    {
        _connected = false;
        _port = null;
        _lastError = null;
    }

    public bool IsTransferMode() => false;

    public IReadOnlyList<Measurement> FetchMeasurements() =>
        throw new NotImplementedException(
            "PVPM serial protocol is not yet implemented. " +
            "Use the VendorExportDriver instead."
        );
}

File: Solarica/hint_from_chatgpt/pvpm-platform/apps/local-reader/app/drivers/serial_driver.py
from __future__ import annotations
import serial.tools.list_ports

class SerialDirectDriver:
    def __init__(self) -> None:
        self.connected = False
        self.port: str | None = None
        self.last_error: str | None = None

    def detect(self) -> dict:
        return {
            "connected": self.connected,
            "mode": "serial",
            "port": self.port,
            "deviceModel": None,
            "deviceSerial": None,
            "firmwareVersion": None,
            "transferModeRequired": True,
            "transferModeDetected": False,
            "lastError": self.last_error,
        }

    def list_ports(self) -> list[dict]:
        return [
            {"name": port.device, "description": port.description}
            for port in serial.tools.list_ports.comports()
        ]

    def connect(self, port: str) -> dict:
        self.connected = True
        self.port = port
        self.last_error = "Protocol not implemented yet. Use diagnostics to capture serial traffic."
        return self.detect()

    def disconnect(self) -> None:
        self.connected = False
        self.port = None

    def is_transfer_mode(self) -> bool:
        return False

    def fetch_device_info(self) -> dict:
        return self.detect()

    def fetch_measurements(self) -> list[dict]:
        raise NotImplementedError("Serial protocol not implemented yet.")
